### DISCUSSION TITLE
API Data Mismatch: Incorrect Party Listings for Docket 68118928

### DISCUSSION THREAD

**Discussion by jasminegump** [freelawproject/courtlistener#4016] [Q&A]
While reviewing the docket for [United States v. The State of West Virginia] (https://www.courtlistener.com/docket/5541058/united-states-v-the-state-of-west-virginia/), I've encountered a discrepancy between the number of parties listed in different API endpoints.

The API's parties endpoint (https://www.courtlistener.com/api/rest/v3/parties/?docket__id=5541058) for this docket should list 4 defendants and 4 plaintiffs if consistent with what's displayed on the docket page. However, the parties endpoint is showing an unexpectedly large number of entries, nearly including every person in the database. 

Could someone investigate why the parties API is returning a complete or near-complete list of individuals from the database for this specific docket? Assistance in resolving this discrepancy would be greatly appreciated.

**Comment by mlissner** [comment]
The way parties link to dockets is one of the more complicated relationships in our data. I'm seeing a couple things that are weird here: 

1. The API is returning more than 20 results without pagination. 
2. The API is returning more than a few results. The docket returns just a few.

Weird stuff. I'm hopeful that this is just us using the API incorrectly. We'll continue investigating.

Thanks for raising this.

**Comment by mlissner** [comment]
OK, I understand what it's doing, but I don't have an opinion about whether it's right yet. The API is not returning more than 20 results. It's returning 3 parties. The problem is that one of the parties is the United States, and for each party, the API is returning all of the attorneys that ever represented that party (not just the ones that repped the party *in that case*). As you might expect, a LOT of attorneys have repped the U.S. over the years....

**Comment by mlissner** [comment]
Thinking aloud here, I think it might just be doing the right thing. The API is designed to return *parties*, and that's precisely what it's doing. It's returning the parties (and everything we know about them) from the case you selected. 

I'm trying to think if there's a smart way that we could make it only return the attorneys in a selected case. Maybe a new URL parameter that works with the `docket` filter:

```
/parties/?docket=xyz&show_all_attys=False
```

Or, perhaps it's better to do that by default when you filter to a docket? But if we do that, the documentation has to explain that a magic thing is happening. My gut is that this is better, but it would be a breaking change. We're doing v4 of the API right now though, so the timing is good for breaking changes.

PS: Doing `docket__id` isn't necessary and makes the query slower. You can just do `docket=xyz`.

**Reply by mlissner** [reply]
I added a new issue to discuss (and follow) solutions to this issue, if you're interested, @jasminegump: https://github.com/freelawproject/courtlistener/issues/4054

**Comment by jasminegump** [comment]
Gotcha! On the non API endpoint (https://www.courtlistener.com/docket/5541058/united-states-v-the-state-of-west-virginia/), how is the US being rendered as a party? How does it know to not list every single attorney in the US party?

**Comment by mlissner** [comment]
Well in that case, the context is obviously the docket, so we filter to just the attorneys for that party on that docket. 

Here, the context is the parties *for that docket*, which means we filter to all parties on that docket, and include all the attorneys for those parties.

**Reply by jasminegump** [reply]
Sorry, I'm still trying to understand how this works using a specific example. For instance, at the docket endpoint (https://www.courtlistener.com/api/rest/v3/dockets/5541058/), the data shows:
"plaintiff": "UNITED STATES OF AMERICA",
"defendant": "THE STATE OF WEST VIRGINIA, ET AL",

Based on this information, do you filter attorneys associated with a specific party? If yes, how is this done? Do you perform a string comparison, or is there another method you use to query this data?

**Reply by jasminegump** [reply]
@mlissner I was wondering if you had a chance to look at my question about? Thank you!

**Reply by mlissner** [reply]
Sorry, I missed your question entirely. So the trick here is to filter to the attorneys and party_types that you care about. So when you see: 

![image](https://github.com/freelawproject/courtlistener/assets/236970/e0bd5e2c-b6ac-41ad-9115-5413ca6aa4d8)

Just do a little filtering to only include attorneys and party_types that have the docket that you want. 

In the screenshot it's a bit annoying because we have URLs for the docket, but I just did https://github.com/freelawproject/courtlistener/pull/4026, which will add `docket_id` and `attorney_id` values to the nested objects. 

Does that clarify?
